### DISCUSSION TITLE
Next: Steps for the rust rewrite

### DISCUSSION THREAD

**Discussion by LDprg** [winapps-org/winapps#57] [General]
This is just a discussion where we specify what next needs to be done or not regarding the rust rewrite. 
Feel free to make suggestions!

**Comment by AkechiShiro** [comment]
Could this be of any use @oskardotglobal @LDprg  ?
https://git.m-labs.hk/M-Labs/wfvm

**Reply by LDprg** [reply]
@AkechiShiro this looks actually really nice for building custom windows images. It would be nice if we could combine this with Github Actions for automatic image builds. 

Also expanding wfvm with layers for custom software or settings seems not that difficult.

The only problem I see is that custom images in general might not be working great with quickemu. I did some testing a while ago with some custom (stripped) images they didn't worked that well. 

Maybe we should consider to use qemu or libvirt directly via executables or rust bindings.

**Reply by oskardotglobal** [reply]
I had no issues with custom images & quickemu
For the images we could probably just use ameliorated playbooks which can be built using the https://git.ameliorated.info/Styris/trusted-uninstaller-cli (although it isn't really documented all that much)

**Reply by oskardotglobal** [reply]
Actually, it seems like this could be perfect in combination with images from ameliorated

**Comment by LDprg** [comment]
I think we should implement the vm management with docker next.

**Comment by LDprg** [comment]
Another thing to do is to implement the autosetup for the shortcuts of the apps.

We should also rework the freerdp arguments to keep compatibility with freerdp3 and freerdp2. Basically something similar to #87 but for the rewrite.

**Comment by purepani** [comment]
Hey, I might be interested in contributing at some point to a rewrite. Is there any information about what the state of the rewrite is in, or which branch it is being worked on?

**Reply by AkechiShiro** [reply]
Hey @purepani,

We have 2 branches for the rewrite but at the moment, we haven't been very active on it : 
- https://github.com/winapps-org/winapps/commits/rewrite/
- https://github.com/winapps-org/winapps/commits/rewrite-freerdp-backend/

**Reply by purepani** [reply]
I remember reading a while back about GPL issues; would it be worth rewriting from scratch in a different project, or does this rust rewrite deal with those licence issues?

**Reply by oskardotglobal** [reply]
The rewrite wouldn't be affected by these issues.

**Reply by LDprg** [reply]
> I remember reading a while back about GPL issues; would it be worth rewriting from scratch in a different project, or does this rust rewrite deal with those licence issues?

That's the reason why the rewrite exists in the first place. The second reason is that shell scripts are hard to maintain compared to rust code.

**Reply by LDprg** [reply]
@purepani As mentioned before the rewrite has it own branch (the "rewrite" one is more up to date).

Basic functionality works, but there is some stuff missing. Here are some of them:
- No automatic desktop links
- The automatic vm management still uses virsh it should be rewritten to use docker
- No automatic app detection
- Configuration is missing some important configs and should get better default values
- possibly more backends like looking glass (especially for wayland users)
- documentation and tutorials on how to use the winapps rewrite (this however will need to wait until other points are completed)

That's all I remember. The desktop links, docker vm management and app detection are the most important ones. 

Feel free to fork, open a pr and ask for feedback.

**Reply by oskardotglobal** [reply]
Actually, the vm management uses quickemu but yes

**Reply by purepani** [reply]
Sounds good. If I have a chance to I'll see if I can get into the project.
